File: src/Azdo.Core/Metrics/Trends.cs
using System.Globalization;

namespace Azdo.Core.Metrics;

/// <summary>The time range during which a sprint tag was active in the snapshot file.</summary>
public sealed class SprintWindow
{
    public string Tag { get; set; } = "";
    public DateTime Start { get; set; }

    /// <summary>Inclusive.</summary>
    public DateTime End { get; set; }

    public SprintWindow() { }

    public SprintWindow(string tag, DateTime start, DateTime end)
    {
        Tag = tag;
        Start = start;
        End = end;
    }
}

/// <summary>One row of the sprint-on-sprint comparison.</summary>
public sealed class TrendRow
{
    public string User { get; set; } = "";
    public List<TrendCell> Cells { get; set; } = new();
}

/// <summary>The four numbers rendered in each sprint's column for one user, plus the overload flag.</summary>
public sealed class TrendCell
{
    public double Points { get; set; }
    public double AvgWIP { get; set; }
    public int StuckCount { get; set; }
    public TimeSpan CycleTime { get; set; }
    public bool OverloadedAnyDay { get; set; }
}

/// <summary>Pure trend windowing and per-user × per-sprint aggregation.</summary>
public static class Trends
{
    private const string BucketActive = "active";
    private const string BucketRFT = "rft";
    private const string BucketClosed = "closed";

    /// <summary>
    /// Returns the time range for <paramref name="tag"/> derived purely from
    /// snapshot rows. ok=false if the tag is never seen.
    /// </summary>
    public static (SprintWindow Window, bool Ok) DeriveSprintWindow(
        IReadOnlyList<Snapshot> snaps, string tag, DateTime now, StateConfig states)
    {
        string earliest = "", latestOpen = "", latestClosed = "", latestAny = "";
        foreach (var s in snaps)
        {
            if (!HasTag(s, tag)) continue;
            if (earliest == "" || string.CompareOrdinal(s.TS, earliest) < 0) earliest = s.TS;
            if (latestAny == "" || string.CompareOrdinal(s.TS, latestAny) > 0) latestAny = s.TS;
            if (states.IsClosed(s.State))
            {
                if (latestClosed == "" || string.CompareOrdinal(s.TS, latestClosed) > 0) latestClosed = s.TS;
            }
            else
            {
                if (latestOpen == "" || string.CompareOrdinal(s.TS, latestOpen) > 0) latestOpen = s.TS;
            }
        }
        if (earliest == "")
            return (new SprintWindow(), false);

        var start = ParseDay(earliest);
        DateTime end;
        if (latestOpen != "" && string.CompareOrdinal(latestOpen, latestClosed) > 0)
            end = now; // still in flight
        else if (latestOpen != "")
            end = ParseDay(latestOpen);
        else
            end = ParseDay(latestAny); // all observations closed

        return (new SprintWindow(tag, start, end), true);
    }

    /// <summary>
    /// Folds snapshot rows into per-user × per-sprint cells. For each window, only
    /// rows whose Tags include the window's Tag AND whose snapshot TS falls inside
    /// [Start, End] contribute.
    /// </summary>
    public static List<TrendRow> TrendAggregate(
        IReadOnlyList<Snapshot> snaps, IReadOnlyList<SprintWindow> windows, Thresholds th, DateTime now)
    {
        var rows = new List<TrendRow>();
        if (windows.Count == 0)
            return rows;

        var users = new HashSet<string>();
        var store = new Dictionary<(string User, int WIdx), Acc>();

        Acc GetAcc(string user, int wIdx)
        {
            var k = (user, wIdx);
            if (!store.TryGetValue(k, out var a))
            {
                a = new Acc();
                store[k] = a;
            }
            return a;
        }

        var states = th.States;

        foreach (var s in snaps)
        {
            if (string.IsNullOrEmpty(s.AssignedTo) || s.AssignedTo == "-")
                continue;
            if (!Snapshots.TryParseDate(s.TS, out var d))
                continue;

            string bucket;
            if (states.IsActive(s.State)) bucket = BucketActive;
            else if (states.IsRFT(s.State)) bucket = BucketRFT;
            else if (states.IsClosed(s.State)) bucket = BucketClosed;
            else continue;

            for (int wIdx = 0; wIdx < windows.Count; wIdx++)
            {
                var w = windows[wIdx];
                if (!HasTag(s, w.Tag)) continue;
                if (d < StripTime(w.Start) || d > StripTime(w.End)) continue;

                users.Add(s.AssignedTo);
                var a = GetAcc(s.AssignedTo, wIdx);

                if (bucket == BucketActive || bucket == BucketRFT)
                {
                    if (!a.StateDays.TryGetValue(s.Id, out var byBucket))
                    {
                        byBucket = new Dictionary<string, HashSet<string>>();
                        a.StateDays[s.Id] = byBucket;
                    }
                    if (!byBucket.TryGetValue(bucket, out var dayset))
                    {
                        dayset = new HashSet<string>();
                        byBucket[bucket] = dayset;
                    }
                    dayset.Add(s.TS);

                    if (!a.DailyWIP.TryGetValue(s.TS, out var ids))
                    {
                        ids = new HashSet<int>();
                        a.DailyWIP[s.TS] = ids;
                    }
                    ids.Add(s.Id);

                    if (bucket == BucketActive)
                    {
                        if (!a.ClosedItemFirst.TryGetValue(s.Id, out var cur) || d < cur)
                            a.ClosedItemFirst[s.Id] = d;
                    }
                }
                else // closed
                {
                    if (!a.ClosedItemDone.TryGetValue(s.Id, out var cur) || d < cur)
                        a.ClosedItemDone[s.Id] = d;
                    a.ClosedPoints[s.Id] = s.Points;
                }
            }
        }

        var names = users.ToList();
        names.Sort(StringComparer.Ordinal);

        foreach (var u in names)
        {
            var row = new TrendRow { User = u, Cells = new List<TrendCell>(windows.Count) };
            for (int wIdx = 0; wIdx < windows.Count; wIdx++)
                row.Cells.Add(new TrendCell());

            for (int wIdx = 0; wIdx < windows.Count; wIdx++)
            {
                if (!store.TryGetValue((u, wIdx), out var a))
                    continue;

                int total = 0, days = 0, peak = 0;
                foreach (var ids in a.DailyWIP.Values)
                {
                    int n = ids.Count;
                    if (n > peak) peak = n;
                    total += n;
                    days++;
                }
                double avg = days > 0 ? (double)total / days : 0.0;

                var sumDur = TimeSpan.Zero;
                int closedN = 0;
                foreach (var kv in a.ClosedItemDone)
                {
                    if (!a.ClosedItemFirst.TryGetValue(kv.Key, out var startD))
                        continue;
                    var dur = kv.Value - startD;
                    if (dur < TimeSpan.Zero) continue;
                    sumDur += dur;
                    closedN++;
                }
                var cy = closedN > 0 ? TimeSpan.FromTicks(sumDur.Ticks / closedN) : TimeSpan.Zero;

                var stuckSet = new HashSet<int>();
                foreach (var kv in a.StateDays)
                {
                    if (th.ActiveStaleDays > 0
                        && kv.Value.TryGetValue(BucketActive, out var ad) && ad.Count > th.ActiveStaleDays)
                        stuckSet.Add(kv.Key);
                    if (th.RFTStaleDays > 0
                        && kv.Value.TryGetValue(BucketRFT, out var rd) && rd.Count > th.RFTStaleDays)
                        stuckSet.Add(kv.Key);
                }

                double points = 0;
                foreach (var p in a.ClosedPoints.Values) points += p;

                row.Cells[wIdx] = new TrendCell
                {
                    Points = points,
                    AvgWIP = avg,
                    StuckCount = stuckSet.Count,
                    CycleTime = cy,
                    OverloadedAnyDay = th.WIPLimit > 0 && peak > th.WIPLimit,
                };
            }
            rows.Add(row);
        }
        return rows;
    }

    internal static bool HasTag(Snapshot s, string tag)
    {
        foreach (var t in s.Tags)
            if (t == tag) return true;
        return false;
    }

    private static DateTime ParseDay(string ts) =>
        DateTime.ParseExact(ts, "yyyy-MM-dd", CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal);

    private static DateTime StripTime(DateTime t) =>
        new(t.Year, t.Month, t.Day, 0, 0, 0, t.Kind);

    private sealed class Acc
    {
        public Dictionary<int, double> ClosedPoints { get; } = new();
        public Dictionary<string, HashSet<int>> DailyWIP { get; } = new();
        public Dictionary<int, Dictionary<string, HashSet<string>>> StateDays { get; } = new();
        public Dictionary<int, DateTime> ClosedItemFirst { get; } = new();
        public Dictionary<int, DateTime> ClosedItemDone { get; } = new();
    }
}

File: reference/internal/metrics/trends.go
package metrics

import (
	"sort"
	"time"
)

// Internal bucket keys used by TrendAggregate's accumulator maps. Decoupled
// from the configured state names so the per-item-state tracking doesn't have
// to deal with case folding inside the inner loop.
const (
	bucketActive = "active"
	bucketRFT    = "rft"
	bucketClosed = "closed"
)

// SprintWindow is the time range during which a sprint tag was active in the
// snapshot file. End is inclusive — Trends counts a day fully if any item
// carried the tag on that day.
type SprintWindow struct {
	Tag   string
	Start time.Time
	End   time.Time
}

// TrendRow is one row of the sprint-on-sprint comparison: a user and one
// TrendCell per sprint window passed to TrendAggregate.
type TrendRow struct {
	User  string
	Cells []TrendCell
}

// TrendCell carries the four numbers rendered in each sprint's column for one
// user, plus the overload flag.
type TrendCell struct {
	Points           float64       // story points on items closed inside the window
	AvgWIP           float64       // average daily in-flight count (Active+RFT)
	StuckCount       int           // distinct items that exceeded a stale threshold on at least one day in the window
	CycleTime        time.Duration // average (first-Active → first-Closed) for items closed in the window
	OverloadedAnyDay bool          // peak daily in-flight > WIPLimit at any point in the window
}

// DeriveSprintWindow returns the time range for `tag` derived purely from
// snapshot rows.
//   - Start = earliest TS the tag appears in any state.
//   - End:
//   - If there are open (non-Closed) observations later than the latest
//     Closed observation, the sprint is still in flight → End = now.
//   - Otherwise the sprint has wrapped → End = latest non-Closed TS (the
//     last day the sprint had items in flight). If every observation is
//     Closed, fall back to the latest TS seen.
//
// `states` carries the configured Closed name so wrapped-sprint detection
// works for teams that don't call the terminal state "Closed". ok=false if
// the tag is never seen in the snapshot.
func DeriveSprintWindow(snaps []Snapshot, tag string, now time.Time, states StateConfig) (SprintWindow, bool) {
	var earliest, latestOpen, latestClosed, latestAny string
	for _, s := range snaps {
		if !hasTag(s, tag) {
			continue
		}
		if earliest == "" || s.TS < earliest {
			earliest = s.TS
		}
		if latestAny == "" || s.TS > latestAny {
			latestAny = s.TS
		}
		if states.IsClosed(s.State) {
			if latestClosed == "" || s.TS > latestClosed {
				latestClosed = s.TS
			}
		} else {
			if latestOpen == "" || s.TS > latestOpen {
				latestOpen = s.TS
			}
		}
	}
	if earliest == "" {
		return SprintWindow{}, false
	}

	start, _ := time.Parse("2006-01-02", earliest)
	var end time.Time
	switch {
	case latestOpen != "" && latestOpen > latestClosed:
		// Sprint still in flight (open observations newer than any Closed).
		end = now
	case latestOpen != "":
		end, _ = time.Parse("2006-01-02", latestOpen)
	default:
		// All observations Closed — pin to the last seen row.
		end, _ = time.Parse("2006-01-02", latestAny)
	}
	return SprintWindow{Tag: tag, Start: start, End: end}, true
}

// TrendAggregate folds snapshot rows into per-user × per-sprint cells. Pure.
//
// For each window, only rows whose Tags include the window's Tag AND whose
// snapshot TS falls inside [Start, End] contribute. Rows belonging to other
// sprints are ignored.
func TrendAggregate(snaps []Snapshot, windows []SprintWindow, th Thresholds, now time.Time) []TrendRow {
	if len(windows) == 0 {
		return nil
	}

	// users[name] = true means we've seen at least one row for that user
	// in any window — so we render their row.
	users := make(map[string]bool)
	// per (user, windowIdx) accumulators
	type acc struct {
		closedPoints    map[int]float64                        // item → points, summed once at finalize
		dailyWIP        map[string]map[int]struct{}            // day → set of item IDs in-flight
		stateDays       map[int]map[string]map[string]struct{} // item → state → set of TS
		closedItemFirst map[int]time.Time                      // item → first-Active observation
		closedItemDone  map[int]time.Time                      // item → first-Closed observation
	}
	cells := make([][]*acc, len(windows))
	for i := range cells {
		cells[i] = nil
	}
	// index keyed by user name for fast row lookup per window
	type key struct {
		user string
		wIdx int
	}
	store := make(map[key]*acc)
	getAcc := func(user string, wIdx int) *acc {
		k := key{user, wIdx}
		a := store[k]
		if a == nil {
			a = &acc{
				closedPoints:    make(map[int]float64),
				dailyWIP:        make(map[string]map[int]struct{}),
				stateDays:       make(map[int]map[string]map[string]struct{}),
				closedItemFirst: make(map[int]time.Time),
				closedItemDone:  make(map[int]time.Time),
			}
			store[k] = a
		}
		return a
	}

	states := th.States

	for _, s := range snaps {
		// Skip unassigned rows — they're not a developer signal and would
		// otherwise produce a no-owner row in the grid.
		if s.AssignedTo == "" || s.AssignedTo == "-" {
			continue
		}
		d, err := time.Parse("2006-01-02", s.TS)
		if err != nil {
			continue
		}

		// Map the snapshot's raw state into a bucket sentinel (or skip).
		var bucket string
		switch {
		case states.IsActive(s.State):
			bucket = bucketActive
		case states.IsRFT(s.State):
			bucket = bucketRFT
		case states.IsClosed(s.State):
			bucket = bucketClosed
		default:
			continue // not one of the three tracked buckets (e.g. New)
		}

		for wIdx, w := range windows {
			if !hasTag(s, w.Tag) {
				continue
			}
			if d.Before(stripTime(w.Start)) || d.After(stripTime(w.End)) {
				continue
			}
			users[s.AssignedTo] = true
			a := getAcc(s.AssignedTo, wIdx)

			// Track observed days per (item, bucket) so stuck-count is computed
			// from observations alone, no reliance on StateSince.
			if bucket == bucketActive || bucket == bucketRFT {
				if a.stateDays[s.ID] == nil {
					a.stateDays[s.ID] = make(map[string]map[string]struct{})
				}
				if a.stateDays[s.ID][bucket] == nil {
					a.stateDays[s.ID][bucket] = make(map[string]struct{})
				}
				a.stateDays[s.ID][bucket][s.TS] = struct{}{}
			}

			switch bucket {
			case bucketActive, bucketRFT:
				if a.dailyWIP[s.TS] == nil {
					a.dailyWIP[s.TS] = make(map[int]struct{})
				}
				a.dailyWIP[s.TS][s.ID] = struct{}{}

				if bucket == bucketActive {
					if cur, ok := a.closedItemFirst[s.ID]; !ok || d.Before(cur) {
						a.closedItemFirst[s.ID] = d
					}
				}
			case bucketClosed:
				if cur, ok := a.closedItemDone[s.ID]; !ok || d.Before(cur) {
					a.closedItemDone[s.ID] = d
				}
				// Record the item's points keyed by ID — multiple daily Closed
				// rows for the same item overwrite the same key, so the total
				// is summed once per distinct item at finalize.
				a.closedPoints[s.ID] = s.Points
			}
		}
	}

	// Build rows, one per user, in alphabetical order.
	names := make([]string, 0, len(users))
	for u := range users {
		names = append(names, u)
	}
	sort.Strings(names)

	rows := make([]TrendRow, len(names))
	for i, u := range names {
		row := TrendRow{User: u, Cells: make([]TrendCell, len(windows))}
		for wIdx, w := range windows {
			a := store[key{u, wIdx}]
			if a == nil {
				continue
			}

			// Average WIP across the window's days (count days w/ activity).
			total, days := 0, 0
			peak := 0
			for _, ids := range a.dailyWIP {
				n := len(ids)
				if n > peak {
					peak = n
				}
				total += n
				days++
			}
			avg := 0.0
			if days > 0 {
				avg = float64(total) / float64(days)
			}

			// Cycle time: average across items that have both a first-Active
			// and a first-Closed observation inside this window.
			var sumDur time.Duration
			closedN := 0
			for id, done := range a.closedItemDone {
				start, ok := a.closedItemFirst[id]
				if !ok {
					continue
				}
				dur := done.Sub(start)
				if dur < 0 {
					continue
				}
				sumDur += dur
				closedN++
			}
			cy := time.Duration(0)
			if closedN > 0 {
				cy = sumDur / time.Duration(closedN)
			}

			// Stuck count: distinct items observed in the same in-flight
			// state for MORE than the threshold's day count within this window.
			stuckSet := make(map[int]struct{})
			for id, byBucket := range a.stateDays {
				if th.ActiveStaleDays > 0 {
					if days := byBucket[bucketActive]; len(days) > th.ActiveStaleDays {
						stuckSet[id] = struct{}{}
					}
				}
				if th.RFTStaleDays > 0 {
					if days := byBucket[bucketRFT]; len(days) > th.RFTStaleDays {
						stuckSet[id] = struct{}{}
					}
				}
			}

			var points float64
			for _, p := range a.closedPoints {
				points += p
			}

			row.Cells[wIdx] = TrendCell{
				Points:           points,
				AvgWIP:           avg,
				StuckCount:       len(stuckSet),
				CycleTime:        cy,
				OverloadedAnyDay: th.WIPLimit > 0 && peak > th.WIPLimit,
			}
			_ = w // window is captured in row order, not stored per-cell
		}
		rows[i] = row
	}
	return rows
}

// hasTag reports whether the snapshot row carries the given tag.
func hasTag(s Snapshot, tag string) bool {
	for _, t := range s.Tags {
		if t == tag {
			return true
		}
	}
	return false
}

